Analysis/PIDTools/src/Compute_dEdxProcessor.cc

```diff
@@ -104,136 +177,305 @@ void Compute_dEdxProcessor::processEvent( LCEvent * evt ) {
+      r=1.0/trk->getOmega();
+      l=r*2.*asin(0.5*dxy/r);
+      over=sqrt(l*l + pow(cross_1[2]-cross_2[2],2));
+
+      totover += over;
+      cover++;
+      dEdxVec3.push_back(hit->getEDep()/over);
     }
+
+    streamlog_out(DEBUG3) << "1: " << length << "  2: " << path << "  3: " << over << "  EDep: " << hit->getEDep() << std::endl;
+    if (length/over>1.5 || over/length>1.5) streamlog_out(DEBUG2) << "  " << hitX << " " << hitY << " " << hitZ << " " << sqrt(hitX*hitX + hitY*hitY) << std::endl;
   }
+  streamlog_out(DEBUG4) << "1: " << totlength << "  2: " << totpath  << " of " << cpath << "  3: " << totover << " of " << cover << std::endl;
 
   //second, get truncated points vector
   //changing vectors in descending order
-  float tmpe=0.0;
-  for(int iH=0;iH<itpc;iH++){
-    for(int jH=iH+1;jH<itpc;jH++){
-      if(depe[iH]<depe[jH]){
-        tmpe=depe[iH];
-        depe[iH]=depe[jH];
-        depe[jH]=tmpe;
-      }
-    }
-  }
+//  float tmpe=0.0;
+//  for(int iH=0;iH<itpc;iH++){
+//    for(int jH=iH+1;jH<itpc;jH++){
+//      if(depe[iH]<depe[jH]){
+//        tmpe=depe[iH];
+//        depe[iH]=depe[jH];
+//        depe[jH]=tmpe;
+//      }
+//    }
+//  }
+  std::sort(dEdxVec1.begin(),dEdxVec1.end());
+  std::sort(dEdxVec2.begin(),dEdxVec2.end());
+  std::sort(dEdxVec3.begin(),dEdxVec3.end());
 
   //check max. and min energy to be truncated(min 8% and max 30%) (aleph: min 8% max 40%)
-  int maxi=std::max((int)(0.30*itpc)-1,0);
-  int mini=std::min((int)(0.92*itpc)-1,itpc-1);
-  float maxe=depe[maxi];
-  float mine=depe[mini];
+//  int maxi=std::max((int)(0.30*itpc)-1,0);
+//  int mini=std::min((int)(0.92*itpc)-1,itpc-1);
+//  float maxe=depe[maxi];
+//  float mine=depe[mini];
   //float unum=(float)(mini-maxi+1);
 
-  //std::cout << "check max and min: " << maxe << " " << mine << std::endl;
-  //
-  //    Reset
-  //
-  double dedx = 0.0;
+  double dedx1=0, dedx2=0, dedx3=0;  // total dEdx values for the 3 dx strategies
+  int nTruncate=0, nTruncate1=0, nTruncate2=0, nTruncate3=0;  // number of remaining track hits after truncation
+
   //int biH=-1;
-  for(int iH=0;iH<itpc;iH++){
-    //TrackerHitExtended * hitExt = hitVec[iH];
-    //TrackerHit * hit = hitExt->getTrackerHit();
-    if(float(depe[iH])>mine && float(depe[iH])<maxe){   //need to check here
-      //if(float(depe[iH])<maxe){   //need to check here
-      dedx+=depe[iH];
-      nTruncate++;
-      //biH=iH;
-    }
+//  for(int iH=0;iH<itpc;iH++){
+//    //TrackerHitExtended * hitExt = hitVec[iH];
+//    //TrackerHit * hit = hitExt->getTrackerHit();
+//    if(float(depe[iH])>mine && float(depe[iH])<maxe){   //need to check here
+//      //if(float(depe[iH])<maxe){   //need to check here
+//      dedx+=depe[iH];
+//      nTruncate++;
+//      //biH=iH;
+//    }
+//  }
+
+  for (int idE=dEdxVec1.size()*_lowerTrunFrac; idE<dEdxVec1.size()*(1-_upperTrunFrac); idE++) {dedx1 += dEdxVec1[idE]; nTruncate1++;}
```

Comment:
Can you use your option _dxStrategy to make a choice on the dEdX calculation and avoid to compute it 3 times. If I choose an option, I don't want to spend time on calculation for the 2 others options I don't want.

Reply:
done